internal/eventstore/handler/crdb/projection_lock.go

```diff
@@ -0,0 +1,62 @@
+package crdb
+
+import (
+	"context"
+	"time"
+
+	"github.com/caos/zitadel/internal/errors"
+)
+
+const (
+	lockStmtFormat = "INSERT INTO %[1]s" +
+		" (locker_id, locked_until, projection_name) VALUES ($1, now()+$2::INTERVAL, $3)" +
+		" ON CONFLICT (projection_name)" +
+		" DO UPDATE SET locker_id = $1, locked_until = now()+$2::INTERVAL" +
+		" WHERE %[1]s.projection_name = $3 AND (%[1]s.locker_id = $1 OR %[1]s.locked_until < now())"
+)
+
+func (h *StatementHandler) Lock(ctx context.Context, lockDuration time.Duration) <-chan error {
+	errs := make(chan error)
+	go h.handleLock(ctx, errs, lockDuration)
+	return errs
+}
+
+func (h *StatementHandler) handleLock(ctx context.Context, errs chan error, lockDuration time.Duration) {
+	renewLock := time.NewTimer(0)
+	for {
+		select {
+		case <-renewLock.C:
+			errs <- h.renewLock(ctx, lockDuration)
+			//refresh the lock 500ms before it times out. 500ms should be enough for one transaction
+			renewLock.Reset(lockDuration - (500 * time.Millisecond))
```

Comment:
is there a check for the lockduration to be at least 500ms?

Reply:
implemented in the start method of projection handler